src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -895,21 +895,10 @@ InstanceKlass* SystemDictionary::resolve_class_from_stream(
   assert(class_name == nullptr || class_name == h_name, "name mismatch");
 
   // Add class just loaded
-  // If a class loader supports parallel classloading, handle parallel define requests.
-  // find_or_define_instance_class may return a different InstanceKlass,
-  // in which case the old k would be deallocated
-  if (is_parallelCapable(class_loader)) {
-    k = find_or_define_instance_class(h_name, class_loader, k, CHECK_NULL);
-  } else {
-    define_instance_class(k, class_loader, THREAD);
-
-    // If defining the class throws an exception register 'k' for cleanup.
-    if (HAS_PENDING_EXCEPTION) {
-      assert(k != nullptr, "Must have an instance klass here!");
-      loader_data->add_to_deallocate_list(k);
-      return nullptr;
-    }
-  }
+  // If a class loader supports parallel classloading with AllowParallelDefineClass for parallel-capable class non-boot
+  // class loaders, handle parallel define requests. In this case, find_or_define_instance_class may return a different
+  // InstanceKlass, in which case the old k would be deallocated
+  k = find_or_define_instance_class(h_name, class_loader, k, CHECK_NULL);
```

Comment:
Now that you call this for parallel-capable and non-parallel-capable loaders the comments preceding the definition of `find_or_define_instance_class` no longer make complete sense.

Reply:
This comment is still true.  non bootstrap parallel-capable class loaders might return a different k.

Reply:
Edit. yes this comment repeats what find_or_define_instance_class already says later on, so doesn't really add anything here.

Reply:
What would make more sense is find_or_define_instance_class was renamed define_instance_class (which actually handled the AllowParallelDefine logic without bothering the caller about it).

Reply:
maybe not.  It does return a 'k'.

Reply:
My point is that this:
```
// If a class loader supports parallel classloading handle parallel define requests.
// find_or_define_instance_class may return a different InstanceKlass
InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_name, Handle class_loader,
                                                               InstanceKlass* k, TRAPS) {
```
makes it seem like this method is only for parallel-capable loaders when that is no longer true. The same goes for the large comment block ahead of `find_or_define_helper`. These methods do not document how they work for non-parallel-capable loaders.